### PR TITLE
feat: push node statistics to an influxDB database

### PR DESCRIPTION
**Pull request by alexander-camuto**
- logs node statistics to an influxDB v2 endpoint specified in environment variables

### CODE REVIEW

File: node/server.go [hunk 1]
```diff
@@ -342,6 +343,16 @@ func Run(ctx context.Context, opts Options) error {
 		node: nd,
 	}
 
+	if opts.UseInflux {
```
Comment (alexander-camuto) [inline node/server.go:346]: @tchardin this is where the initial check-in happens -- we just need to run a similar operation whenever a retrieval occurs and calculate stats on that retrieval accordingly

**Comment by alexander-camuto** [comment]
@tchardin as we discussed we need to include stats on discovery time. To calculate this on the provider end we need the OG client request to be timestamped. When a provider then serves that request it can also calculate the discovery time + retrieval time and push it to the DB.

**Comment by tchardin** [comment]
@alexander-camuto this looks good, I would simplify a few things:
- I don't think we actually need to change influxdb params while the node is running do we? I would just set the metrics variables via the flag (which also works as env variables)
So you could do:
```go
// in metrics
type Config struct {
    InfluxURL string
}
// in start.go
var metricsConfig metrics.Config

fs.StringVar(&metricsConfig.InfluxURL, "influxdb_url", "", "database url for recording metrics")
```
Then the pop node options would have a field for metricsConfig, if all the right environment variables are available upon start you can pass a pointer with them to the Options struct. If the metrics field is not nil, it means metrics should be activated. This avoids passing a -use-influx flag, you can just pass your environment variables like you usually do:
`INFLUXDB_URL=...`

```go
// in popn.go
type Options struct {
    metrics *metricsConfig
}
```
Let me know if that makes sense. We can start with your SendMessage method but I don't think it would be bad to have the influx db as part of a metrics interface on the node:
```go
type MetricsRecorder interface {
    Record(string, interface{})
}

type node struct {
    metrics MetricsRecorder
}
```
Which means that any implementation can then implement that interface and send any serialized data to whatever implementation we decide to use:

```go
// implements MetricsRecorder
type InfluxDBClient {
    client influxdb.Client
    org ...etc
}

func (idbc *InfluxDBClient) Record(name string, value interface{}) {
 p := influxdb.NewPoint(name, value)

  // get non-blocking write client
  writeAPI := idbc.client.WriteAPI(idbc.Org, idbc.Bucket)

  // write point asynchronously
  writeAPI.WritePoint(p)
  // Flush writes
  writeAPI.Flush()
}
```

Then in pop node:
```go
node.metrics = metrics.NewInfluxDBClient(options.metrics)
```

You feel me?

**Comment by alexander-camuto** [comment]
@tchardin latest push should address those comments.

File: node/popn.go [hunk 2]
```diff
@@ -124,6 +127,11 @@ type Options struct {
 	CancelFunc context.CancelFunc
 }
 
+type MetricsRecorder interface {
+    Record(string, map[string]string, map[string]interface{})
+		URL() string
```
Comment (tchardin) [inline node/popn.go:132]: What's that for?

File: cmd/pop/cli/start.go [hunk 3]
```diff
@@ -24,6 +25,7 @@ import (
 // PopConfig is the json config object we generate with the init command
 type PopConfig struct {
 	temp         bool
+	influxdb     bool
```
Comment (tchardin) [inline cmd/pop/cli/start.go:28]: You can remove that

File: metrics/influxdb_client.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  client := influxdb.NewClient(params.InfluxURL, params.InfluxToken)
+
+  return &InfluxDBClient{client, params.InfluxURL, params.Org, params.Bucket}
+
+}
+
+func GetInfluxParams() *Config {
+  addr := os.Getenv(EnvInfluxDBURL)
+	if addr == "" {
+		return &Config{}
```
Comment (tchardin) [inline metrics/influxdb_client.go:70]: All of these you can return `nil`

File: node/popn.go [hunk 5]
```diff
@@ -290,6 +299,10 @@ func New(ctx context.Context, opts Options) (*node, error) {
 		return nil, err
 	}
 
+	if *opts.Metrics != (metrics.Config{}) {
```
Comment (tchardin) [inline node/popn.go:302]:
```suggestion
	if opts.Metrics != nil {
```

File: node/server.go [hunk 6]
```diff
@@ -338,6 +339,13 @@ func Run(ctx context.Context, opts Options) error {
 		fmt.Printf("==> Connected to Filecoin RPC at %s\n", opts.FilEndpoint)
 	}
 
+	if reflect.TypeOf(nd.metrics) != nil {
```
Comment (tchardin) [inline node/server.go:342]:
```suggestion
	if nd.metrics != nil {
```

Comment (tchardin) [reply node/server.go:342]: What you could do is create an empty Metrics interface so you don't need to check every time if nd.metrics is nil

Comment (tchardin) [reply node/server.go:342]: So you'd do:
```go
type NullMetrics struct {}

func (*NullMetrics) Record(...) {}

etc
```
Then when you initialize it you could have a contructor `NewMetrics` that either returns an InfluxDB if the params are there of a NullMetrics

File: node/popn.go (hunk 2 above)
Comment (alexander-camuto) [reply node/popn.go:132]: just a handy thing for logging what database URL we're hitting up

File: metrics/influxdb_client.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  return idbc.InfluxURL
+}
+
+type NullMetrics struct {}
+
+func (*NullMetrics) Record(name string, tag map[string]string, value map[string]interface{}) {}
+
+func (*NullMetrics) URL() string { return "" }
+
+func NewMetrics(params *Config) MetricsRecorder {
```
Comment (tchardin) [inline metrics/influxdb_client.go:58]: You can actually make this:
```suggestion
func New(params *Config) MetricsRecorder {
```

File: node/popn.go [hunk 8]
```diff
@@ -290,6 +295,8 @@ func New(ctx context.Context, opts Options) (*node, error) {
 		return nil, err
 	}
 
+	nd.metrics = metrics.NewMetrics(opts.Metrics)
```
Comment (tchardin) [inline node/popn.go:298]:
```suggestion
	nd.metrics = metrics.New(opts.Metrics)
```

File: metrics/influxdb_client.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+type Config struct {
+  InfluxURL string
+  InfluxToken string
+  Org string
+  Bucket string
+}
+
+type MetricsRecorder interface {
+  Record(string, map[string]string, map[string]interface{})
+  URL() string
```
Comment (tchardin) [inline metrics/influxdb_client.go:25]: Not a fan of this method cause it seems kind of random on a metrics recorder interface but if you absolutely want it sure

Comment (alexander-camuto) [reply metrics/influxdb_client.go:25]: def want it atm

Comment (alexander-camuto) [reply metrics/influxdb_client.go:25]: just want it for logging purposes atm

**Review by tchardin — APPROVED** [review]

**Pull request merged**